google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractLazyInitializer.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+/**
+ * Generic {@link AbstractLazyInitializer} for any heavy-weight object that might throw an exception
+ * during initialization. The underlying object is initialized at most once.
+ */
+public abstract class AbstractLazyInitializer<T> {
```

Comment:
Could we use composition instead of inheritance (pass in the initialiser function instead of overriding it)?

Reply:
That would be possible, but as we have to support Java 7, that would mean that we would have to use `com.google.cloud.base.Function` in the public interface. I think that it is more idiomatic to use inheritance than composition for Java 7 (once we drop Java 7 support, I would agree that using composition would be a lot nicer in this case).